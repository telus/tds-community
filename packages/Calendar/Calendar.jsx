import React, { Component } from 'react'
import PropTypes from 'prop-types'
import momentPropTypes from 'react-moment-proptypes'

import 'react-dates/initialize'
import SingleDatePicker from 'react-dates/lib/components/SingleDatePicker'
import DayPickerSingleDateController from 'react-dates/lib/components/DayPickerSingleDateController'
import 'react-dates/lib/css/_datepicker.css'

import safeRest from '@tds/shared-safe-rest'

import DecorativeIcon from '@tds/core-decorative-icon'

import { CalendarContainer, LabelText } from './styles'

/**
 * Calendar component for choosing available dates.
 * @version ./package.json
 */

class Calendar extends Component {
  state = {
    focused: false,
  }

  onFocusChange = ({ focused }) => {
    this.setState({ focused })
  }
  render() {
    const { id, date, onDateChange, isDayDisabled, inline, label, ...props } = this.props
    const { className, style, ...propsWithoutStyling } = safeRest(props)
    const DatePickerVariant = inline ? DayPickerSingleDateController : SingleDatePicker

    /* Determine daySize based on window.outerWidth and `inline` */
    const getResponsiveDaySize = () => {
      let responsiveDaySize = 40
      if (window) {
        if (window.outerWidth >= 432) {
          responsiveDaySize = inline ? 60 : 48
        } else {
          responsiveDaySize = window.outerWidth >= 376 ? 40 : 32
        }
      } else {
        return responsiveDaySize
      }
      return responsiveDaySize
    }

    const getIcon = type => <DecorativeIcon symbol={type} size={16} />

    /* eslint-disable jsx-a11y/label-has-for */
    return (
      <CalendarContainer {...safeRest(propsWithoutStyling)}>
        <label htmlFor={id} />
        <LabelText>{label}</LabelText>
        <DatePickerVariant
          id={id}
          date={date}
          onDateChange={onDateChange}
          focused={this.state.focused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          hideKeyboardShortcutsPanel={true}
          displayFormat="DD / MM / YYYY"
          placeholder="DD / MM / YYYY"
          isDayBlocked={isDayDisabled}
          keepOpenOnDateSelect={false}
          daySize={getResponsiveDaySize()}
          navPrev={getIcon('leftChevron')}
          navNext={getIcon('chevron')}
        />
      </CalendarContainer>
    )
  }
}

Calendar.propTypes = {
  /** A unique identifier */
  id: PropTypes.string.isRequired,

  /** A Moment instance representing the currently selected date, i.e. `moment()` */
  date: momentPropTypes.momentObj,

  /** Event triggered every time a new date is clicked on
  @param {Moment} date The new date that was selected */
  onDateChange: PropTypes.func,

  /** A function determining whether a given date should be disabled
  @param {Moment} date The date to optionally disable
  @returns {bool} */
  isDayDisabled: PropTypes.func,

  /** The field label to be displayed above the calendar */
  label: PropTypes.string.isRequired,

  /** A flag determining if the calendar picker is standalone or an input with overlay  */
  inline: PropTypes.bool,
}

Calendar.defaultProps = {
  isDayDisabled: undefined,
  date: undefined,
  onDateChange: () => {},
  inline: false,
}

export default Calendar
