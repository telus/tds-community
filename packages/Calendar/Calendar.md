### Usage Criteria

- `Calendar` component can be used for appointment booking or other scenarios where users need to pick a date.
- Use `inline` prop to render inline calendar picker which does not contain input field.
- Prop `inline` is `false` by default.

```jsx
initialState = { date: moment() }
;<Calendar
  id="appointment_calendar"
  label="Book an appointment"
  date={state.date}
  onDateChange={date => setState({ date })}
/>
```

### Inline Calendar Picker

```jsx
initialState = { date: moment() }
;<Calendar
  inline
  label="Book an appointment"
  date={state.date}
  onDateChange={date => setState({ date })}
/>
```
