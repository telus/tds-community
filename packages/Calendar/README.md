# TDS Community: Calendar

A calendar component that can be used for appointment booking or other scenarios where users need to pick a date.
