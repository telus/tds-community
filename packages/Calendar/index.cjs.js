require('./dist/index.css')
const Calendar = require('./dist/index.cjs')

module.exports = Calendar
