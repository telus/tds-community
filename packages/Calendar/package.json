{
  "name": "@tds/community-calendar",
  "version": "1.0.0-preview",
  "description": "",
  "main": "index.cjs.js",
  "module": "index.es.js",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/telus/tds-community.git"
  },
  "publishConfig": {
    "access": "public"
  },
  "author": "TELUS digital",
  "engines": {
    "node": ">=8"
  },
  "bugs": {
    "url": "https://github.com/telus/tds-community/issues"
  },
  "homepage": "http://tds.telus.com",
  "peerDependencies": {
    "react": "^16.8.2",
    "react-dom": "^16.8.2",
    "styled-components": "^4.1.3"
  },
  "dependencies": {
    "@tds/core-colours": "^1.1.1",
    "@tds/core-decorative-icon": "^1.2.0",
    "@tds/shared-typography": "1.2.1",
    "prop-types": "^15.6.2",
    "react-dates": "^19.0.3",
    "react-moment-proptypes": "^1.6.0"
  },
  "devDependencies": {
    "@tds/shared-safe-rest": "^1.0.0",
    "@tds/core-responsive": "^1.1.0",
    "moment": "^2.24.0"
  }
}
