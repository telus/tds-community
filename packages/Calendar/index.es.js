import './dist/index.css'
import Calendar from './dist/index.es'

export default Calendar
