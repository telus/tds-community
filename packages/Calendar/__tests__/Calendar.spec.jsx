import React from 'react'
import { shallow, mount } from 'enzyme'
import moment from 'moment'
import Calendar from '../Calendar'

describe('Calendar', () => {
  const defaultProps = {
    id: 'appointment_calendar',
    label: 'Book an appointment',
    date: moment(1608924597000),
  }
  const doShallow = (props = defaultProps) => shallow(<Calendar {...props} />)
  const doMount = (props = defaultProps) => mount(<Calendar {...props} />)

  it('renders', () => {
    const calendar = doShallow()

    expect(calendar).toMatchSnapshot()
  })

  it('opens the calendar when input is in focus', () => {
    const calendar = doMount()

    calendar.find('.DateInput_input').simulate('focus')

    expect(calendar.find('.SingleDatePicker_picker')).toHaveLength(1)
  })

  it('passes additional attributes to the element', () => {
    const calendar = doShallow({
      ...defaultProps,
      'data-some-attr': 'some value',
    })

    expect(calendar).toHaveProp('data-some-attr', 'some value')
  })

  it('does not allow custom CSS', () => {
    const calendar = doShallow({
      ...defaultProps,
      className: 'my-custom-class',
      style: { color: 'hotpink' },
    })

    expect(calendar).not.toHaveProp('className', 'my-custom-class')
    expect(calendar).not.toHaveProp('style')
  })
})
