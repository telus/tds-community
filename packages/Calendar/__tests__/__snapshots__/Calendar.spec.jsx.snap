// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar renders 1`] = `
<styles__CalendarContainer>
  <label
    htmlFor="appointment_calendar"
  >
    <styles__LabelText>
      Book an appointment
    </styles__LabelText>
  </label>
  <withStyles(SingleDatePicker)
    anchorDirection="left"
    appendToBody={false}
    block={false}
    calendarInfoPosition="bottom"
    customCloseIcon={null}
    customInputIcon={null}
    date={"2020-12-25T19:29:57.000Z"}
    daySize={48}
    disableScroll={false}
    disabled={false}
    displayFormat="DD / MM / YYYY"
    enableOutsideDays={false}
    firstDayOfWeek={null}
    focused={false}
    hideKeyboardShortcutsPanel={true}
    horizontalMargin={0}
    horizontalMonthPadding={13}
    id="appointment_calendar"
    initialVisibleMonth={null}
    inputIconPosition="before"
    isDayBlocked={[Function]}
    isDayHighlighted={[Function]}
    isOutsideRange={[Function]}
    isRTL={false}
    keepFocusOnInput={false}
    keepOpenOnDateSelect={false}
    monthFormat="MMMM YYYY"
    navNext={
      <DecorativeIcon
        size={16}
        symbol="chevron"
      />
    }
    navPrev={
      <DecorativeIcon
        size={16}
        symbol="leftChevron"
      />
    }
    noBorder={false}
    numberOfMonths={1}
    onClose={[Function]}
    onDateChange={[Function]}
    onFocusChange={[Function]}
    onNextMonthClick={[Function]}
    onPrevMonthClick={[Function]}
    openDirection="down"
    orientation="horizontal"
    phrases={
      Object {
        "calendarLabel": "Calendar",
        "chooseAvailableDate": [Function],
        "clearDate": "Clear Date",
        "closeDatePicker": "Close",
        "dateIsSelected": [Function],
        "dateIsUnavailable": [Function],
        "enterKey": "Enter key",
        "escape": "Escape key",
        "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
        "homeEnd": "Home and end keys",
        "jumpToNextMonth": "Move forward to switch to the next month.",
        "jumpToPrevMonth": "Move backward to switch to the previous month.",
        "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
  select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
        "keyboardShortcuts": "Keyboard Shortcuts",
        "leftArrowRightArrow": "Right and left arrow keys",
        "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
        "moveFocusByOneMonth": "Switch months.",
        "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
        "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
        "openThisPanel": "Open this panel.",
        "pageUpPageDown": "page up and page down keys",
        "questionMark": "Question mark",
        "returnFocusToInput": "Return to the date input field.",
        "selectFocusedDate": "Select the date in focus.",
        "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
        "upArrowDownArrow": "up and down arrow keys",
      }
    }
    placeholder="DD / MM / YYYY"
    readOnly={false}
    regular={false}
    renderCalendarInfo={null}
    renderDayContents={null}
    renderMonthElement={null}
    renderMonthText={null}
    reopenPickerOnClearDate={false}
    required={false}
    screenReaderInputMessage=""
    showClearDate={false}
    showDefaultInputIcon={false}
    small={false}
    verticalHeight={null}
    verticalSpacing={22}
    weekDayFormat="dd"
    withFullScreenPortal={false}
    withPortal={false}
  />
</styles__CalendarContainer>
`;
